Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+
+            GRefPtr<GVariant> arguments(g_variant_new("(ssha{sv})", "", _("Print Web Page"), fdIndex, &options));
+
+            auto* priv = printOperation->priv;
+            g_dbus_proxy_call_with_unix_fd_list(priv->portalProxy.get(), "Print", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, fdList.get(), nullptr,
+                [](GObject* proxy, GAsyncResult* result, gpointer userData) {
+                    auto printOperation = adoptGRef(static_cast<WebKitPrintOperation*>(userData));
+                    GUniqueOutPtr<GError> error;
+                    GRefPtr<GVariant> returnValue = adoptGRef(g_dbus_proxy_call_finish(G_DBUS_PROXY(proxy), result, &error.outPtr()));
+                    if (error)
+                        webkitPrintOperationFailed(printOperation.get(), WTFMove(error));
+                    else
+                        webkitPrintOperationFinished(printOperation.get());
+                }, printOperation.leakRef());
+        }, callbackData, nullptr);
+    });
+}
+
+static void webkitPrintOperationPreparePrint(WebKitPrintOperation* printOperation)
+{
+    auto* priv = printOperation->priv;
+
+    RELEASE_ASSERT(priv->portalProxy);
+
+    auto* connection = g_dbus_proxy_get_connection(priv->portalProxy.get());
+    auto uniqueName = String::fromUTF8(g_dbus_connection_get_unique_name(connection));
+    auto senderName = makeStringByReplacingAll(uniqueName.substring(1), '.', '_');
+    auto token = makeString("WebKitGTK", weakRandomNumber<uint32_t>());
+    auto requestPath = makeString("/org/freedesktop/portal/desktop/request/", senderName, "/", token);
+
+    RELEASE_ASSERT(!priv->signalId);
+
+    priv->signalId = g_dbus_connection_signal_subscribe(connection, "org.freedesktop.portal.Desktop", "org.freedesktop.portal.Request",
+        "Response", requestPath.ascii().data(), nullptr, G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
+        [](GDBusConnection* connection, const char*, const char*, const char*, const char*, GVariant* parameters, gpointer userData) {
+            GRefPtr<WebKitPrintOperation> printOperation(static_cast<WebKitPrintOperation*>(userData));
+            auto* priv = printOperation->priv;
+            GRefPtr<GVariant> options;
+            uint32_t response;
+
+            RELEASE_ASSERT(priv->signalId);
+            g_dbus_connection_signal_unsubscribe(connection, priv->signalId);
+            priv->signalId = 0;
+
+            g_variant_get(parameters, "(u@a{sv})", &response, &options.outPtr());
+
+            if (response) {
+                priv->preparePrintResponse = PreparePrintResponse(WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL);
+                return;
+            }
+
+            GRefPtr<GVariant> printSettingsVariant(g_variant_lookup_value(options.get(), "settings", G_VARIANT_TYPE_VARDICT));
+            GRefPtr<GVariant> pageSetupVariant(g_variant_lookup_value(options.get(), "page-setup", G_VARIANT_TYPE_VARDICT));
```

Comment:
Does `g_variant_lookup_value` return a floating ref?

Reply:
No. This was leaking, I think. Hopefully fixed now.